helix-dap/src/client.rs

```diff
@@ -0,0 +1,702 @@
+    pub data: Option<Value>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct StoppedEventBody {
+    pub reason: String,
+    pub description: Option<String>,
+    pub thread_id: Option<usize>,
+    pub preserve_focus_hint: Option<bool>,
+    pub text: Option<String>,
+    pub all_threads_stopped: Option<bool>,
+    pub hit_breakpoint_ids: Option<Vec<usize>>,
+}
+
+#[derive(Debug)]
+pub struct Client {
+    id: usize,
+    _process: Option<Child>,
+    server_tx: UnboundedSender<Request>,
+    request_counter: AtomicU64,
+    capabilities: Option<DebuggerCapabilities>,
+    awaited_events: Arc<Mutex<HashMap<String, Sender<Event>>>>,
+}
+
+impl Client {
+    pub fn streams(
+        rx: Box<dyn AsyncBufRead + Unpin + Send>,
+        tx: Box<dyn AsyncWrite + Unpin + Send>,
+        id: usize,
+        process: Option<Child>,
+    ) -> Result<Self> {
+        let (server_rx, server_tx) = Transport::start(rx, tx, id);
+
+        let client = Self {
+            id,
+            _process: process,
+            server_tx,
+            request_counter: AtomicU64::new(0),
+            capabilities: None,
+            awaited_events: Arc::new(Mutex::new(HashMap::default())),
+        };
+
+        tokio::spawn(Self::recv(Arc::clone(&client.awaited_events), server_rx));
+
+        Ok(client)
+    }
+
+    pub async fn tcp(addr: std::net::SocketAddr, id: usize) -> Result<Self> {
+        let stream = TcpStream::connect(addr).await?;
+        let (rx, tx) = stream.into_split();
+        Self::streams(Box::new(BufReader::new(rx)), Box::new(tx), id, None)
+    }
+
+    pub fn stdio(cmd: &str, args: Vec<&str>, id: usize) -> Result<Self> {
+        let process = Command::new(cmd)
+            .args(args)
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            // make sure the process is reaped on drop
+            .kill_on_drop(true)
+            .spawn();
+
+        let mut process = process?;
+
+        // TODO: do we need bufreader/writer here? or do we use async wrappers on unblock?
+        let writer = BufWriter::new(process.stdin.take().expect("Failed to open stdin"));
+        let reader = BufReader::new(process.stdout.take().expect("Failed to open stdout"));
+
+        Self::streams(
+            Box::new(BufReader::new(reader)),
+            Box::new(writer),
+            id,
+            Some(process),
+        )
+    }
+
+    async fn get_port() -> Option<u16> {
+        Some(
+            tokio::net::TcpListener::bind(SocketAddr::new(
+                IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
+                0,
+            ))
+            .await
+            .ok()?
+            .local_addr()
+            .ok()?
+            .port(),
+        )
+    }
+
+    pub async fn tcp_process(
+        cmd: &str,
+        args: Vec<&str>,
+        port_format: &str,
+        id: usize,
+    ) -> Result<Self> {
+        let port = Self::get_port().await.unwrap();
```

Comment:
This might have race condition in really rare situation (like lack of ports which might increase the probability):
1. We acquired ephemeral port from OS
2. Another process took it
3. Debug adapter fails to start because port is acquired by other process.